# AI Threshold Analysis 

1. ImageNet 
2. CUB

In [1]:
import numpy as np
import pandas as pd
from tqdm import tqdm
import pickle
import random

# ImageNet
##  Load summary fo AI Performance

In [2]:
random.seed(42)

with open('../data/Classification-Summary-ImageNet.pickle', 'rb') as f:
  classification_summary = pickle.load(f)

In [3]:
classification_summary.keys()

dict_keys(['all_gts', 'resnet_wnid', 'knn_wnid', 'emd_wnid', 'chm_wnid', 'resnet_conf', 'knn_conf', 'emd_conf', 'chm_conf', 'resnet_wnid_real', 'knn_wnid_real', 'emd_wnid_real', 'chm_wnid_real', 'resnet_real_conf', 'knn__real_conf', 'emd__real_conf', 'chm__real_conf'])

### Unpacking

In [4]:
real_gt_real = classification_summary['all_gts']

resnet_wnid_real = classification_summary['resnet_wnid_real']
knn_wnid_real    = classification_summary['knn_wnid_real']   
emd_wnid_real    = classification_summary['emd_wnid_real']   
chm_wnid_real    = classification_summary['chm_wnid_real']   

IsM1Correct_real = classification_summary['resnet_wnid_real'] 
IsM2Correct_real = classification_summary['knn_wnid_real']    
IsM3Correct_real = classification_summary['emd_wnid_real']    
IsM4Correct_real = classification_summary['chm_wnid_real']    

M1Conf_real = classification_summary['resnet_real_conf']  
M2Conf_real = classification_summary['knn__real_conf']    
M3Conf_real = classification_summary['emd__real_conf']    
M4Conf_real = classification_summary['chm__real_conf']    

In [ ]:
MethodName = ['ResNet', 'KNN', 'EMD', 'CHM']
Methods    = [IsM1Correct_real,IsM2Correct_real,IsM3Correct_real,IsM4Correct_real]
Confs      = [M1Conf_real,M2Conf_real,M3Conf_real,M4Conf_real]

TAnalysis = {}

for N, M, C in zip(MethodName, Methods, Confs):
  TAnalysis[N] = {}
  for T in tqdm(np.arange(0., 1.05, 0.05)):
    TAnalysis[N][round(T, 2)] = np.average([x[0] == (x[1]>T) for x in zip(M, C)])

In [6]:
def split_data_44(a, b):
  zipped = list(zip(a, b))
  random.shuffle(zipped)
  ra, rb = zip(*zipped)
  return ra[:2000], rb[:2000],  ra[2000:], rb[2000:]

In [7]:
MethodName = ['ResNet', 'KNN', 'EMD', 'CHM']

IsM1Correct_T, M1Conf_T, IsM1Correct_V, M1Conf_V = split_data_44(IsM1Correct_real, M1Conf_real)
IsM2Correct_T, M2Conf_T, IsM2Correct_V, M2Conf_V = split_data_44(IsM2Correct_real, M2Conf_real)
IsM3Correct_T, M3Conf_T, IsM3Correct_V, M3Conf_V = split_data_44(IsM3Correct_real, M3Conf_real)
IsM4Correct_T, M4Conf_T, IsM4Correct_V, M4Conf_V = split_data_44(IsM4Correct_real, M4Conf_real)


Methods_T = [IsM1Correct_T, IsM2Correct_T, IsM3Correct_T, IsM4Correct_T]
Confs_T   = [M1Conf_T, M2Conf_T, M3Conf_T, M4Conf_T]

TAnalysis_T = {}

for N, M, C in zip(MethodName, Methods_T, Confs_T):
  TAnalysis_T[N] = {}
  for T in tqdm(np.arange(0., 1.05, 0.05)):
    TAnalysis_T[N][round(T, 2)] = np.average([x[0] == (x[1]>T) for x in zip(M, C)])

100%|██████████| 21/21 [00:00<00:00, 508.07it/s]


In [8]:
Methods_V = [IsM1Correct_V, IsM2Correct_V, IsM3Correct_V, IsM4Correct_V]
Confs_V   = [M1Conf_V, M2Conf_V, M3Conf_V, M4Conf_V]

TAnalysis_V = {}

for N, M, C in zip(MethodName, Methods_V, Confs_V):
  TAnalysis_V[N] = {}
  for T in tqdm(np.arange(0., 1.05, 0.05)):
    TAnalysis_V[N][round(T, 2)] = np.average([x[0] == (x[1]>T) for x in zip(M, C)])

100%|██████████| 21/21 [00:00<00:00, 24.21it/s]


In [9]:
pd.DataFrame(TAnalysis).T.style.background_gradient(axis=1)

In [10]:
pd.DataFrame(TAnalysis_T).T.style.background_gradient(axis=1)

In [11]:
pd.DataFrame(TAnalysis_V).T.style.background_gradient(axis=1)

In [12]:
# Full 44K
pd.DataFrame(TAnalysis).T.style.highlight_max(color = 'yellow', axis = 1)

In [13]:
# 2K Fine-tuning
pd.DataFrame(TAnalysis_T).T.style.highlight_max(color = 'yellow', axis = 1)

In [26]:
# 42K Test
pd.DataFrame(TAnalysis_V).T.loc[:, 0.35:0.45]*100

0.35       0.40       0.45
ResNet  85.864134  86.050349  86.102206
KNN     85.866491  85.946634  85.569489
EMD     85.913634  85.821705  85.420988
CHM     85.479917  85.635489  85.362059

# RESULTS for ImageNet

* ResNet  ``86.050349``
* KNN     ``85.946634``
* EMD     `85.913634`
* CHM     ``85.362059``

# CUB
##  Load summary fo AI Performance

In [15]:
random.seed(42)

with open('../data/Classification-Summary-CUB.pickle', 'rb') as f:
  classification_summary_cub = pickle.load(f)

In [16]:
classification_summary_cub.keys()

dict_keys(['GT', 'resnet_pred', 'knn_pred', 'emd_pred', 'chm_pred', 'resnet_wnid', 'knn_wnid', 'emd_wnid', 'chm_wnid', 'resnet_conf', 'knn_conf', 'emd_conf', 'chm_conf'])

### Unpacking 

In [17]:
GT_CUB     = classification_summary_cub['GT']

M1Prediction_CUB = classification_summary_cub['resnet_pred']
M2Prediction_CUB = classification_summary_cub['knn_pred']   
M3Prediction_CUB = classification_summary_cub['emd_pred']   
M4Prediction_CUB = classification_summary_cub['chm_pred']   
   
IsM1Correct_CUB = classification_summary_cub['resnet_wnid']
IsM2Correct_CUB = classification_summary_cub['knn_wnid']   
IsM3Correct_CUB = classification_summary_cub['emd_wnid']   
IsM4Correct_CUB = classification_summary_cub['chm_wnid']   

M1Conf_CUB = classification_summary_cub['resnet_conf'] 
M2Conf_CUB = classification_summary_cub['knn_conf']    
M3Conf_CUB = classification_summary_cub['emd_conf']    
M4Conf_CUB = classification_summary_cub['chm_conf']    

In [18]:
MethodName_CUB = ['ResNet', 'KNN', 'EMD', 'CHM']
Methods_CUB    = [IsM1Correct_CUB, IsM2Correct_CUB, IsM3Correct_CUB, IsM4Correct_CUB]
Confs_CUB      = [M1Conf_CUB, M2Conf_CUB, M3Conf_CUB,M4Conf_CUB]

TAnalysis_CUB = {}

for N, M, C in zip(MethodName_CUB, Methods_CUB, Confs_CUB):
  TAnalysis_CUB[N] = {}
  for T in tqdm(np.arange(0., 1.05, 0.05)):
    TAnalysis_CUB[N][round(T, 2)] = np.average([x[0] == (x[1]>T) for x in zip(M, C)])

100%|██████████| 21/21 [00:00<00:00, 176.03it/s]


## Fine-tune on 1K, Test on 4K

In [19]:
def split_data_1k_4k(a, b):
  zipped = list(zip(a, b))
  random.shuffle(zipped)
  ra, rb = zip(*zipped)
  return ra[:1000], rb[:1000],  ra[1000:], rb[1000:]

In [20]:
MethodName_CUB = ['ResNet', 'KNN', 'EMD', 'CHM']

IsM1Correct_T_CUB, M1Conf_T_CUB, IsM1Correct_V_CUB, M1Conf_V_CUB = split_data_1k_4k(IsM1Correct_CUB, M1Conf_CUB)
IsM2Correct_T_CUB, M2Conf_T_CUB, IsM2Correct_V_CUB, M2Conf_V_CUB = split_data_1k_4k(IsM2Correct_CUB, M2Conf_CUB)
IsM3Correct_T_CUB, M3Conf_T_CUB, IsM3Correct_V_CUB, M3Conf_V_CUB = split_data_1k_4k(IsM3Correct_CUB, M3Conf_CUB)
IsM4Correct_T_CUB, M4Conf_T_CUB, IsM4Correct_V_CUB, M4Conf_V_CUB = split_data_1k_4k(IsM4Correct_CUB, M4Conf_CUB)

Methods_T_CUB = [IsM1Correct_T_CUB, IsM2Correct_T_CUB, IsM3Correct_T_CUB, IsM4Correct_T_CUB]
Confs_T_CUB   = [M1Conf_T_CUB, M2Conf_T_CUB, M3Conf_T_CUB, M4Conf_T_CUB]

TAnalysis_T_CUB = {}

for N, M, C in zip(MethodName_CUB, Methods_T_CUB, Confs_T_CUB):
  TAnalysis_T_CUB[N] = {}
  for T in tqdm(np.arange(0., 1.05, 0.05)):
    TAnalysis_T_CUB[N][round(T, 2)] = np.average([x[0] == (x[1]>T) for x in zip(M, C)])

100%|██████████| 21/21 [00:00<00:00, 974.41it/s]


In [21]:
Methods_V_CUB = [IsM1Correct_V_CUB, IsM2Correct_V_CUB, IsM3Correct_V_CUB, IsM4Correct_V_CUB]
Confs_V_CUB   = [M1Conf_V_CUB, M2Conf_V_CUB, M3Conf_V_CUB, M4Conf_V_CUB]

TAnalysis_V_CUB = {}

for N, M, C in zip(MethodName_CUB, Methods_V_CUB, Confs_V_CUB):
  TAnalysis_V_CUB[N] = {}
  for T in tqdm(np.arange(0., 1.05, 0.05)):
    TAnalysis_V_CUB[N][round(T, 2)] = np.average([x[0] == (x[1]>T) for x in zip(M, C)])

100%|██████████| 21/21 [00:00<00:00, 209.66it/s]


In [22]:
# Full 5K
pd.DataFrame(TAnalysis_CUB).T.style.highlight_max(color = 'yellow', axis = 1)

In [23]:
# 1K Fine-tuning
pd.DataFrame(TAnalysis_T_CUB).T.style.highlight_max(color = 'yellow', axis = 1)

In [24]:
# 4K Test
pd.DataFrame(TAnalysis_V_CUB).T.loc[:, 0.4:0.5]*100

0.40       0.45       0.50
ResNet  86.858573  86.942011  87.108886
KNN     87.400918  87.359199  85.690446
EMD     86.879433  86.211932  84.668335
CHM     85.481852  84.355444  82.165207

# RESULTS for CUB

* ResNet  ``87.108886``
* KNN     ``87.400918``
* EMD     ``86.879433``
* CHM     ``85.481852``